### ISSUE TITLE
[analyzer] [extension types] Method/setter clash diagnostic is confusing

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#53719]
This is the analyzer specific issue for https://github.com/dart-lang/sdk/issues/53717, which has further details.

The analyzer reports the following error for the example in #53717:

```
Class 'E2' can't define method 'm' and have field 'E1.m' with the same name.
```

This is confusing because there are no instance variable declarations named `m` anywhere in this program. It would probably be helpful to report something like "Class 'E2' can't define method 'm' and have a setter 'E1.m' with the same name".

(The language specification documents would say that the name of the setter is `m=` rather than `m`, but presumably the error messages just consider setters to have names like `m` in general, without the `=`.)

**Comment by bwilkerson** [comment]
Yes, diagnostic messages do not include the `=` because users don't see the `=` anywhere in their code.

The rewrite lgtm.

**Comment by scheglov** [comment]
In the analyzer explicit getters and setters declare implicit fields.

**Comment by bwilkerson** [comment]
But in messages to users we should always refer to the explicit getter or setter. Referring to an implicit field is confusing because users don't think of there being an implicit field.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/337641

See also https://github.com/dart-lang/co19/issues/2390

**Comment by eernstg** [comment]
Just to clarify: The language team did support the proposal in https://github.com/dart-lang/language/pull/3470, and https://dart-review.googlesource.com/c/sdk/+/337641 implements the new rules.

(This means that this issue has been reinterpreted from "improve the diagnostic given when we have this kind of method/setter conflict" to "use the new rules based on the notion of 'precludes' to eliminate this kind of method/setter conflict". I've edited the original post in order to indicate this change.)

**Comment by eernstg** [comment]
@scheglov, I adjusted the labels based on the fact that the language has changed and this issue is associated with the implementation of the new rules. Please re-adjust the labels if there's a better way to do it.

**Comment by scheglov** [comment]
@eernstg is this valid code? Do we filter out precluded declarations first, or do we find the unique inherited declaration (report an error if is not unique) and see that it is precluded?
```dart
extension type E1(int it) {
  void foo() {}
}

extension type E2(int it) {
  int get foo => 0;
}

extension type E3(int it) implements E1, E2 {
  set foo(_) {}
}
```

**Comment by eernstg** [comment]
That's a good question!

The current wording in the feature specification makes it an error to have `implements E1, E2` on an extension type that doesn't declare a member named `foo`, and there's no help from having the setter `foo=` because that's not taken into account at that location in the spec.

However, the rule about 'precludes' clearly specifies that `E3` _has_ the getter named `foo`, and not the method.

This means that the original motivation for making the conflict in the superinterface an error on `E3` is gone.

See https://github.com/dart-lang/language/pull/3486.
